src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,138 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        validateIndex(index);
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = head;
+            newNode.prev = null;
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            newNode.next = getNodeByIndex(index);
+            newNode.prev = newNode.next.prev;
+            newNode.next.prev = newNode;
+            newNode.prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> setNode = getNodeByIndex(index);
+        T nodeValue = setNode.value;
+        setNode.value = value;
+        return nodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        T nodeValue = currentNode.value;
+        unlinkNode(currentNode, index);
+        return nodeValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (get(i) == object || object != null && object.equals(get(i))) {
```

Comment:
we should iterate over the nodes and search for the correct node and then pass this node to unlinkNode() method